src/guard.md

```diff
@@ -41,6 +45,7 @@ iff
 if
   (CALLER_ID == Owner) or (CALLED_ID == ACCT_ID) or (Authority == 0)
```

Comment:
Two things I notice here:
 - I think you want "CALLER_ID", not "CALLED_ID", in the second condition
 - If the first two conditions are false but the third is true (so the overall condition is true), `isAuthorized` will return false, reverting the call--so the the "pass" case cannot be accepted even if the typo is fixed:
https://github.com/dapphub/ds-auth/blob/master/src/auth.sol#L63

Reply:
Thank you for catching that typo! 

for the second point, I thought what I was saying here is:
Limit all possible outcomes (with `if`) to one of those three conditions.  But only expect it to not revert with the first two conditions as described in the `iff`.  Since `Authority == 0` is not listed there, it should be a failing case where it would be accepted if it reverts.

Reply:
ok, updating for the typo still gives me accepting proofs for pass/fail on `setOwner` but now the fail proof only reverts because of the VCallValue not because of `Authority == 0` which is unexpected

Reply:
That's not quite how "if" and "iff" work.

"if (condition)" says: if (condition) is true, then the specified behavior occurs
So if you write "if (A or B or C)", so long as any one of A, B, or C is true, your "pass" spec is expected to be accepted. `if` conditions get applied unchanged to any fail specs.

"iff (condition)" says: if (condition) is true, then the specified behavior occurs; also, if (condition) is false, the function reverts
So if you write "iff (A or B)", then the pass spec is expected to be accepted when at least one of A or B is true, and the fail spec is expected to be accepted when A and B are both false (i.e. !(A or B) is true).

So if you write:
iff (A or B)
if (A or B or C)
You're pass case is: specified behavior occurs whenever (A or B) == true and (A or B or C) == true
--> this will not be accepted because truth of either A or B makes both the if and iff true, but C can still be false, and a false C causes reversion, so there's a always a reverting branch in the pass spec when the conditions are written this way

Reply:
Right that is what I wanted/expected but the `Authority == 0` applies in the fail spec.

So the `if` says the following will always be true for both the `pass` and the `fail` spec:
`CALLER_ID == Owner` (Cond A)
`OR`
`CALLER_ID == ACCT_ID` (Cond B)
`OR`
`Authority == 0` (Cond C)

Then the `iff` divides the world into expected passing and expected failing. So we expect 
Cond A
`OR`
Cond B
To give us passing proofs (i.e. in the world of `(A or B or C) == true` where `(A or B) == true` we should pass. But in that world, if `(A or B) != true`, then `(A or B or C) == true` must resolve to `C == true` and that should revert, i.e. give a passing fail proof.

Reply:
Hmmm...looking at the generated K spec in out/specs might clarify how Klab is interpreting the conditions. It does seem like there should be an `Authority == 0` branch in a passing fail spec.

Reply:
going to drop some more debug output here for tracking/comments:

Reply:
for `DSAuth_setOwner_pass_rough`
```act
iff
  VCallValue == 0
  (CALLER_ID == Owner) or ((CALLER_ID == ACCT_ID) and (CALLER_ID =/= Owner))

if
  Authority == 0
```

gives the following k rule:

```
andBool (#rangeAddress(ABI_usr)
  andBool (#rangeAddress(Owner)
  andBool (#rangeAddress(Authority)
  andBool ((#sizeWordStack(CD) <=Int 1250000000)
  andBool ((#notPrecompileAddress(Owner))
  andBool ((#notPrecompileAddress(Authority))
  andBool ((Authority ==Int 0)
  andBool (VGas >=Int 3000000
  andBool ((VCallValue ==Int 0)
  andBool (((CALLER_ID ==Int Owner) orBool ((CALLER_ID ==Int ACCT_ID) andBool (CALLER_ID =/=Int Owner))))
```

and causes `prove` to `reject`. When I add a k debug flag I get the following output:

```
java.lang.AssertionError
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplifyImpl(ConjunctiveFormula.java:442)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:434)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:397)
	at org.kframework.backend.java.kil.ConstrainedTerm.expandPatterns(ConstrainedTerm.java:119)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue$ProcessProofRules.getEvaluatedConstraint(InitializeRewriter.java:391)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue$ProcessProofRules.evaluateRule(InitializeRewriter.java:365)
	at org.kframework.backend.java.kil.Definition.addKoreRule(Definition.java:232)
	at org.kframework.backend.java.kil.Definition.lambda$addKoreRules$9(Definition.java:215)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.kframework.backend.java.kil.Definition.addKoreRules(Definition.java:216)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue$ProcessProofRules.<init>(InitializeRewriter.java:356)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue.prove(InitializeRewriter.java:217)
	at org.kframework.kprove.KProve.run(KProve.java:60)
	at org.kframework.kprove.KProveFrontEnd.run(KProveFrontEnd.java:92)
	at org.kframework.main.FrontEnd.main(FrontEnd.java:59)
	at org.kframework.main.Main.runApplication(Main.java:119)
	at org.kframework.main.Main.runApplication(Main.java:108)
	at org.kframework.main.Main.main(Main.java:54)
[Error] Internal: Uncaught exception thrown of type AssertionError (null)
```

Reply:
for `DSAuth_setOwner_fail_rough`

```act
iff
  VCallValue == 0
  (CALLER_ID == Owner) or ((CALLER_ID == ACCT_ID) and (CALLER_ID =/= Owner))

if
  Authority == 0
```

gives the following k rule:

```
andBool (#rangeAddress(ABI_usr)
  andBool (#rangeAddress(Owner)
  andBool (#rangeAddress(Authority)
  andBool ((#sizeWordStack(CD) <=Int 1250000000)
  andBool ((#notPrecompileAddress(Owner))
  andBool ((#notPrecompileAddress(Authority))
  andBool ((Authority ==Int 0)
  andBool (VGas >=Int 3000000))))))))
  andBool notBool (
    ((VCallValue:Int ==Int 0)
    andBool (((CALLER_ID:Int ==Int Owner:Int) orBool ((CALLER_ID:Int ==Int ACCT_ID:Int) andBool (CALLER_ID:Int =/=Int Owner:Int)))))
  )
```

and causes prove to reject. When I add a k debug flag I get the following output:

```
java.lang.AssertionError
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplifyImpl(ConjunctiveFormula.java:442)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:434)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:397)
	at org.kframework.backend.java.kil.ConstrainedTerm.expandPatterns(ConstrainedTerm.java:119)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue$ProcessProofRules.getEvaluatedConstraint(InitializeRewriter.java:391)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue$ProcessProofRules.evaluateRule(InitializeRewriter.java:365)
	at org.kframework.backend.java.kil.Definition.addKoreRule(Definition.java:232)
	at org.kframework.backend.java.kil.Definition.lambda$addKoreRules$9(Definition.java:215)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:395)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.kframework.backend.java.kil.Definition.addKoreRules(Definition.java:216)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue$ProcessProofRules.<init>(InitializeRewriter.java:356)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue.prove(InitializeRewriter.java:217)
	at org.kframework.kprove.KProve.run(KProve.java:60)
	at org.kframework.kprove.KProveFrontEnd.run(KProveFrontEnd.java:92)
	at org.kframework.main.FrontEnd.main(FrontEnd.java:59)
	at org.kframework.main.Main.runApplication(Main.java:119)
	at org.kframework.main.Main.runApplication(Main.java:108)
	at org.kframework.main.Main.main(Main.java:54)
[Error] Internal: Uncaught exception thrown of type AssertionError (null)
```

Reply:
for `DSAuth_setOwner_pass_rough`

```act
iff
  VCallValue == 0
  (CALLER_ID == Owner)

if
  (CALLER_ID == Owner) or (Authority == 0)
```

gives the following k rule:

```
 andBool (#rangeAddress(ABI_usr)
  andBool (#rangeAddress(Owner)
  andBool (#rangeAddress(Authority)
  andBool ((#sizeWordStack(CD) <=Int 1250000000)
  andBool ((#notPrecompileAddress(Owner))
  andBool ((#notPrecompileAddress(Authority))
  andBool (((CALLER_ID ==Int Owner) orBool (Authority ==Int 0))
  andBool (VGas >=Int 3000000
  andBool ((VCallValue ==Int 0)
  andBool (((CALLER_ID ==Int Owner)))
```

and causes `prove` to `accept`. With the following two branches:

```
  └ 0 ✓   ((Owner ==K ACCT_ID) ==K false)
  └ 1 ✓   (ACCT_ID ==K Owner)
```

Reply:
for `DSAuth_setOwner_fail_rough`

```act
iff
  VCallValue == 0
  (CALLER_ID == Owner)

if
  (CALLER_ID == Owner) or (Authority == 0)
```

gives the following k rule:

```
andBool (#rangeAddress(ABI_usr)
  andBool (#rangeAddress(Owner)
  andBool (#rangeAddress(Authority)
  andBool ((#sizeWordStack(CD) <=Int 1250000000)
  andBool ((#notPrecompileAddress(Owner))
  andBool ((#notPrecompileAddress(Authority))
  andBool (((CALLER_ID ==Int Owner) orBool (Authority ==Int 0))
  andBool (VGas >=Int 3000000))))))))
  andBool notBool (
    ((VCallValue:Int ==Int 0)
    andBool (((CALLER_ID:Int ==Int Owner:Int))))
  ) 
ensures
FAILURE =/=K EVMC_SUCCESS
```

and causes `prove` to `accept`. With the following branch:

```
root ❌
```

which reverts with VCallValue on the stack (making me believe it is because `VCallValue > 0`)

Reply:
for `DSAuth_setOwner_pass_rough`

```act
iff
  VCallValue == 0
  (CALLER_ID == Owner) or (CALLER_ID == ACCT_ID)

if
  (CALLER_ID == Owner) or (CALLER_ID == ACCT_ID) or (Authority == 0)
```

gives the following k rule:

```
andBool (#rangeAddress(ABI_usr)
  andBool (#rangeAddress(Owner)
  andBool (#rangeAddress(Authority)
  andBool ((#sizeWordStack(CD) <=Int 1250000000)
  andBool ((#notPrecompileAddress(Owner))
  andBool ((#notPrecompileAddress(Authority))
  andBool (((CALLER_ID ==Int Owner) orBool (CALLER_ID ==Int ACCT_ID) orBool (Authority ==Int 0))
  andBool (VGas >=Int 3000000
  andBool ((VCallValue ==Int 0)
  andBool (((CALLER_ID ==Int Owner) orBool (CALLER_ID ==Int ACCT_ID)))
```

and causes `prove` to `accept`. With the following two branches:

```
  ├ 0 ✓   ((CALLER_ID ==K ACCT_ID) ==K false)
  └ 1 ✓   (((9 <= CALLER_ID)) AND (((0 < CALLER_ID)) AND (ACCT_ID ==K CALLER_ID)))
```

Reply:
for `DSAuth_setOwner_fail_rough`

```act
iff
  VCallValue == 0
  (CALLER_ID == Owner)

if
  (CALLER_ID == Owner) or (Authority == 0)
```

gives the following k rule:

```andBool (#rangeAddress(ABI_usr)
  andBool (#rangeAddress(Owner)
  andBool (#rangeAddress(Authority)
  andBool ((#sizeWordStack(CD) <=Int 1250000000)
  andBool ((#notPrecompileAddress(Owner))
  andBool ((#notPrecompileAddress(Authority))
  andBool (((CALLER_ID ==Int Owner) orBool (CALLER_ID ==Int ACCT_ID) orBool (Authority ==Int 0))
  andBool (VGas >=Int 3000000))))))))
  andBool notBool (
    ((VCallValue:Int ==Int 0)
    andBool (((CALLER_ID:Int ==Int Owner:Int) orBool (CALLER_ID:Int ==Int ACCT_ID:Int))))
  ) 
ensures
FAILURE =/=K EVMC_SUCCESS
```

and causes `prove` to `accept`. With the following branch:

```
root ❌
```

which reverts with VCallValue on the stack (making me believe it is because `VCallValue > 0`)